### ISSUE TITLE
Drop Ruby 2.4 support

### ISSUE THREAD

**Issue by tenderlove** [brianmario/mysql2#1228]
Ruby core has [dropped all support for Ruby 2.4](https://www.ruby-lang.org/en/news/2020/04/05/support-of-ruby-2-4-has-ended/).  Should we drop it as well?  I know this is kind of cheating, but we have failing builds on Ruby 2.4 and I don't think it's worth our time to fix them if Ruby 2.4 isn't supported anymore.

cc @sodabrew

**Comment by sodabrew** [comment]
My recent hiatus from using Ruby as a daily-driver aside, I think back to how much time I spent making sure that old versions of Ruby continued to work in several gems...

If we did make Ruby 2.5+ required, I would want to do a major version bump. And actually, it's been on my mind for a looong time that the next major should be 1.0. There's nothing that substantially changed in this API for many releases and it would allow Rails to relax/remove its version requirement in the mysql2 adapter to prevent runtime version errors.

The really salient problem I think is also compelling to figure out a policy for is what to do about people who've been posting that they want to backport newer-MySQL and newer-MariaDB support to mysql2 0.3.x so that they can continue to use older versions of Rails with newer versions of MySQL 8.
